cpp/src/arrow/compute/exec/tpch_node.cc

```diff
@@ -0,0 +1,3431 @@
+    *(out + num_padding_zeros + num_digits - 1) = ('0' + x % 10);
+    num_digits -= 1;
+    x /= 10;
+  }
+}
+
+Result<std::shared_ptr<Schema>> SetOutputColumns(
+    const std::vector<std::string>& columns,
+    const std::vector<std::shared_ptr<DataType>>& types,
+    const std::unordered_map<std::string, int>& name_map, std::vector<int>& gen_list) {
+  gen_list.clear();
+  std::vector<std::shared_ptr<Field>> fields;
+  if (columns.empty()) {
+    fields.resize(name_map.size());
+    gen_list.resize(name_map.size());
+    for (auto pair : name_map) {
+      int col_idx = pair.second;
+      fields[col_idx] = field(pair.first, types[col_idx]);
+      gen_list[col_idx] = col_idx;
+    }
+    return schema(std::move(fields));
+  } else {
+    for (const std::string& col : columns) {
+      auto entry = name_map.find(col);
+      if (entry == name_map.end()) return Status::Invalid("Not a valid column name");
+      int col_idx = static_cast<int>(entry->second);
+      fields.push_back(field(col, types[col_idx]));
+      gen_list.push_back(col_idx);
+    }
+    return schema(std::move(fields));
+  }
+}
+
+Result<Datum> RandomVString(random::pcg32_fast& rng, int64_t num_rows, int32_t min_length,
+                            int32_t max_length) {
+  std::uniform_int_distribution<int32_t> length_dist(min_length, max_length);
+  ARROW_ASSIGN_OR_RAISE(std::unique_ptr<Buffer> offset_buff,
+                        AllocateBuffer((num_rows + 1) * sizeof(int32_t)));
+  int32_t* offsets = reinterpret_cast<int32_t*>(offset_buff->mutable_data());
+  offsets[0] = 0;
+  for (int64_t i = 1; i <= num_rows; i++) offsets[i] = offsets[i - 1] + length_dist(rng);
+  ARROW_ASSIGN_OR_RAISE(std::unique_ptr<Buffer> str_buff,
+                        AllocateBuffer(offsets[num_rows]));
+  char* str = reinterpret_cast<char*>(str_buff->mutable_data());
+
+  // Spec says to pick random alphanumeric characters from a set of at least
+  // 64 symbols. Now, let's think critically here: 26 letters in the alphabet,
+  // so 52 total for upper and lower case, and 10 possible digits gives 62
+  // characters...
+  // dbgen solves this by including a space and a comma as well, so we'll
+  // copy that.
+  const char alpha_numerics[65] =
+      "0123456789abcdefghijklmnopqrstuvwxyz ABCDEFGHIJKLMNOPQRSTUVWXYZ,";
+  std::uniform_int_distribution<int> char_dist(0, 63);
+  for (int32_t i = 0; i < offsets[num_rows]; i++) str[i] = alpha_numerics[char_dist(rng)];
+
+  ArrayData ad(utf8(), num_rows, {nullptr, std::move(offset_buff), std::move(str_buff)});
+  return std::move(ad);
+}
+
+void AppendNumber(char*& out, int num_digits, int32_t x) {
+  out += (num_digits - 1);
+  while (x > 0) {
+    *out-- = '0' + (x % 10);
+    x /= 10;
+  }
+  out += (num_digits + 1);
+}
+
+void GeneratePhoneNumber(char* out, random::pcg32_fast& rng, int32_t country) {
+  std::uniform_int_distribution<int32_t> three_digit(100, 999);
+  std::uniform_int_distribution<int32_t> four_digit(1000, 9999);
```

Comment:
Just for the record, TPC-H specifies that phone numbers never have a leading zero?

Reply:
yes, the spec specifies to generate in the range 100-999 and 1000-9999